File: ML/ocmbin.m
classdef ocmbin < bincube   
%  Reader for .bin OCM files
%{  
            Cube:         3d image
            Position:     scan position vector
            Data:         generic additional data
                              e.g.:   fps at scan point; average or time series
                                      extra images
                                      PSI & DAQ settings
%}
    properties
        tifs                                            % references to generated .tif files (for external viewing)
    end
    properties(Hidden = true)
        cube_reduced
        header
        headerParser                                    % header tagStringParser object
        overlap                                         % bool: overlap between adjacent data regions?    
        overbytes                                       % bytes of overlap between adjacent data reions
        io_timings                                      % file i/o timing information
        datasets                                        % names of included datasets; what was the point of this?
    end
    
    methods          
        function load_data(self)
            if ~self.is_loaded
                full_read = tic;            
                self.parseMetadata();           
                self.checkOverlap();

                data_read = tic;
    %             try
                    self.extractData2()                
    %             catch
    %                 warning('Data cannot be loaded')
    %             end

                read_time = toc(data_read);     

                self.descale            

                self.io_timings.writeTime = self.MD.Time.ElapsedTime;
                self.io_timings.dataReadTime = read_time;
                self.io_timings.fullReadTime = toc(full_read);
                self.io_timings.ReadRateMBs = self.filesize_gb / read_time * 1024;
                
                self.is_loaded = true;
            end
        end
        
        function unload_data(self)
            if self.is_loaded
                self.cube = zeros(1,1,1);
                self.cube_reduced = zeros(1,1,1);
                self.position = zeros(1);
                self.data = {};
                
                self.is_loaded = false;
            end
        end
        
        function parseMetadata(self)      
            
            f = fopen(self.path, 'rb');
            fseek(f, 0, 'bof');
%             self.header = fgetl(f); 

            self.header = strcat(fread(f, 8192, '*char')');
            fclose(f);            
            
            if isempty(self.header)
                error('File has no header')
            else
            
                % reads single line of file; the first line in the file is the header
                    % However, what if this is not the case? Also: indented header is easier to read...

                % json parser gets confused when it encounters the header padding.
                
                try % or JSON oops               
                    self.MD = jsondecode(self.header);
                catch err % JSON formatting error or corrupt file
                    self.MD = jsondecode(normalize_json(self.header));
                end

                if ~isfield(self.MD, 'Data') % rename metadate fields in case older file is loaded
                    self.rename_MD_fields() 
                end
                
                self.name = self.MD.Main.File.Name;
                self.description = self.MD.Main.File.Description;

                self.MD.Main.Dir = dir(self.path);
                self.filesize = b2relevant(self.MD.Main.Dir.bytes); 
                self.filesize_gb = b2gb(self.MD.Main.Dir.bytes);
                self.MD.Main.File.FileSizeGB = self.filesize_gb;
                self.MD.Main.Dir.path = strcat(self.MD.Main.Dir.folder, '\', self.MD.Main.Dir.name);
                self.MD.Main.Dir.extension = getExtension(self.MD.Main.Dir.name);
                self.tifs = {};

                [~, sys] = memory;            
                self.memory_limit = b2gb(sys.PhysicalMemory.Available * 0.75);

                % recast data metadata fields from struct array into struct struct

                dsts = {self.MD.Data.Name};
                index = 1:length(dsts);
                self.datasets = dsts(~cellfun('isempty', dsts));
                index = index(~cellfun('isempty', dsts));
                dmd = struct;

                for i = 1:length(index)
                    dmd.(self.datasets{i}) = self.MD.Data(index(i));
                end

                self.MD.Data = dmd;

                for i = 1:length(self.datasets)
                    dataset = self.datasets{i};
                    self.MD.Data.(dataset).BytesAllocated = ...
                        self.MD.Data.(dataset).Position.StopByte - self.MD.Data.(dataset).Position.StartByte;
                    self.MD.Data.(dataset).BytesWritten = ...
                        self.MD.Data.(dataset).Position.LastByte - self.MD.Data.(dataset).Position.StartByte;
                end
            end
        end  
        
        function rename_MD_fields(self)
           oldfields = fieldnames(self.MD);
            targets = {'DataProperties', 'TimingProperties', 'MainProperties', 'CameraAttributes', 'ScanProperties'};
            replace = {'Data', 'Time', 'Main', 'CMOS', 'Scan'};
            md = struct;
            
            for i = 1:length(oldfields)
                match = strfind(oldfields, targets{i});
                index = find(not(cellfun('isempty', match)));
                md.(replace{i}) = self.MD.(oldfields{index});
            end 
            
            self.MD = md;
        end
        
        function [overlap,overbytes] = checkOverlap(self)
            % Checks the parsed metadata for overlap between separate data regions
            
            if isempty(self.datasets)
                self.datasets = fieldnames(self.MD.Data);
            end
                       
            overlap = zeros(1,length(self.datasets));
            overbytes = zeros(1,length(self.datasets));
            allocbytes = zeros(1,length(self.datasets));
            writtenbytes = zeros(1,length(self.datasets));
            
            for i = 1:length(self.datasets)
                overlap(i) = (self.MD.Data.(self.datasets{i}).Position.LastByte ...
                    > self.MD.Data.(self.datasets{i}).Position.StopByte);
                overbytes(i) = self.MD.Data.(self.datasets{i}).Position.LastByte ...
                    - self.MD.Data.(self.datasets{i}).Position.StopByte;
                allocbytes(i) = self.MD.Data.(self.datasets{i}).BytesAllocated;
                writtenbytes(i) = self.MD.Data.(self.datasets{i}).BytesWritten;
            end
            
            self.overlap = overlap;
            self.overbytes = array2table([allocbytes; writtenbytes; overbytes], ...
                'VariableNames', self.datasets, 'RowNames', {'Allocated', 'Written', 'Overlap'});
            
            if any(overlap)
               warning('Some of the data regions overlap. Average delta in bytes: %d', mean(overbytes(overbytes ~= 0)))  
            end
        end 
        
        function extractData2(self, load_mode)
            % Extracts data from the .ocmbin file ~region boundaries specified in the metadata
                % load_mode:    files larger than 75% of available RAM are loaded using memmap                
            switch nargin
                case 1 
                    load_mode = 'read';
            end
            
            fields = fieldnames(self.MD.Data);
            N = length(fields); 
            
            % OCM cube
            start = self.MD.Data.cube.Position.StartByte;
            shape = self.MD.Data.cube.Size';                    
            
            self.cube = self.readRegion(start, shape, load_mode);
            self.cube = permute(self.cube, [2,1,3]);
            i = find(strcmp(fields,'cube'));
            fields{i} = '';
            
            % Scan position vector
            start = self.MD.Data.position.Position.StartByte;
            shape = self.MD.Data.position.Size;                
           
            self.position = self.readRegion(start, shape, load_mode);            
            i = find(strcmp(fields,'position'));
            fields{i} = '';
            
            % Generic data            
            for i = 1:length(fields) % rest of the data
                if ~isempty(fields{i}) % better to 'pop' elements from the cell array...
                    start = self.MD.Data.(fields{i}).Position.StartByte;
                    shape = self.MD.Data.(fields{i}).Size';                
                    
                    self.data.(fields{i}) = self.readRegion(start, shape, load_mode);
                    
                end
            end
            
            % Scan timing data
            start = self.MD.Time.Position.StartByte;
            stop = self.MD.Time.Position.LastByte;
            timing_string = self.readRegion(start, stop-start, load_mode, 'string');
            
            try 
                sss = strsplit(timing_string, '[');
                timing_string = ['[', sss{2}];
                
                has_timing = true;
            catch
                timing_string = '';
                has_timing = false;
            end
            
            if has_timing
                try
                    self.MD.Time.Timings = jsondecode(normalize_json(timing_string));
                    timing_loaded = true;
                catch error
                    % but also catch Scan.Finished == false!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
                    warning(getReport(error, 'extended', 'hyperlinks', 'on'))
                    timing_loaded = false;
                end

                if timing_loaded
                    moves = [self.MD.Time.Timings.MoveTick]'; 
                    starts = [self.MD.Time.Timings.StartTick]'; 
                    stops = [self.MD.Time.Timings.StopTick]';

                    delta_move = starts - moves;
                    delta_scan = stops - starts;
                    delta_total = stops - moves;
                    delta_log = circshift(moves,-1) - stops;

                    self.io_timings.moveTime = median(delta_move)*1e-6;
                    self.io_timings.scanTime = median(delta_scan)*1e-6;
                    self.io_timings.stepTime = median(delta_total)*1e-6;
                    self.io_timings.logTime = median(delta_log(1:end-1))*1e-6;
                end
            end
        end
        
        function I = readRegion(self, start, shape, load_mode, dtype)
            % Reads specified region in file            
            switch nargin
                case [0,1,2]
                    error('Data properties not specified')
                case 3
                    if self.size_GB < self.memory_limit
                        load_mode = 'read';
                    else
                        load_mode = 'memmap';
                    end
                    dtype = self.dtype;
                case 4
                    dtype = self.dtype;
            end
            
            
            
            dimensions = length(shape);
            len = prod(shape);
           
            
            if strcmp(load_mode, 'read')
                 switch dimensions
                        case 1                            
                            if strcmp(dtype, 'string') 
                                f = fopen(self.path, 'r');
                                fseek(f, start, -1);
                                I = fread(f, len, '*char');
                                I = I';
                            else
                                f = fopen(self.path, 'r');
                                fseek(f, start, -1);
                                I = fread(f, len, dtype);
                            end
                            fclose(f);
                        case 2
                            f = fopen(self.path, 'r');
                            fseek(f, start, -1);
                            A = fread(f, len, dtype);
                            I = reshape(A, shape)';
                            fclose(f);
                        case 3
                            f = fopen(self.path, 'r');
                            fseek(f, start, -1);
                            A = fread(f, len, dtype);
                            I = reshape(A, shape);
                            I = permute(I, [2,1,3]);
                            fclose(f);
                        otherwise
                            error('Not implemented; dimensions must be 1-3')
                 end   
                 
%                  if dimensions == 2
%                      I = I'; % transpose 2d arrays (assuming timing -> vertical)
%                  end
            elseif strcmp(load_mode, 'memmap')
                error('Size over 8 GB, to be implemented with memmap')
            end
        end
        
        function descale(self)     
            %%% Divide by MD.Data.(...).factor if needed. %%%            
        end
         
        function ij(self, zrange)
            switch nargin
                case 1
                    zrange = 1:length(self.position);
            end
            
            if ~exist('MIJ', 'class')
                Miji;
            end
            
            if isempty(self.cube_reduced)
                self.cube_reduced = uint16(rescale(self.cube, 0, 2^16));
            end
            
            MIJ.createImage(self.MD.Main.File.Name, self.cube_reduced(:,:,zrange), true);
        end
        
        function export_tif(self, normz, dB, name, path, cube) % would be easier if could see whether it was a name or not
            
            default_name = '';
            [~,~,ext] = fileparts(self.path);
            default_path = strcat(                                                                      ...
                        erase(self.path, ext), '.tif'    ...
                    );
            default_normz = 'volume';
            default_dB = false;
            
            switch nargin
                case 1
                    normz = default_normz;
                    dB = false;
                    name = default_name;
                    path = default_path;
                    cube = self.cube;
                case 2
                    dB = default_dB;
                    name = default_name;
                    path = default_path;
                    cube = self.cube;
                case 3
                    name = '';
                    path = default_path;
                    cube = self.cube;   
                case 4
                    path = appendName(default_path, name);
                    cube = self.cube;
                case 5
                    cube = self.cube;
            end          
            
            switch normz % this could be its own 3D normalization function
                case false
                    pass
                case strcmp(normz, 'volume')
                    cube = normalize(cube);
                case strcmp(normz, 'xy')
                    warning('Not implemented.')
                case strcmp(normz, 'xz')
                    warning('Not implemented.')
                case strcmp(normz, 'yz')
                    warning('Not implemented.')
                case strcmp(normz, 'z')
                    warning('Not implemented.')
            end
            
            switch dB % this could be its own 3D dB function
                case false
                    pass
                case  strcmp(dB, 'volume')
                    cube = dB(cube);
            end
            
            options.overwrite = true; % add option: save file or open saved file
            
            if self.filesize_gb > 4
                options.big = true;
            else
                options.big = false;
            end
                        
            saveastiff(cube, path, options);
            
            if ~isfield(self.MD.Main.File, 'tifs')
                self.tifs = {};
            end
            N = length(self.tifs);
            self.tifs{N+1} = path;
            
        end
        
        function tif(self, varargin)
            
            switch length(varargin)
                case 0
                    arg = 1;
                case 1
                    arg = varargin{1};
                otherwise
                    arg = 0;
            end
                       
            if isnumeric(arg)
                % numeric identifier: index in M.FileProperties.tifs list
                warning('Not yet implemented: open specified .tif file with Fiji or MIJI');
            elseif isstring(arg) || ischar(arg)
                switch arg
                    case ''
                        % select by identifier case: path or name
                        warning('Not yet implemented: open specified .tif file with Fiji or MIJI');
                end                        
            end
            
            if ~arg
                if isempty(self.tifs)
                    self.export_tif(varargin{:});
                    self.tif;
                end
            end
        end
        
        function explore(self)
            winopen(self.MD.Main.File.folder)
        end
    end    
end

File: py/ocmbin.py
import numpy as np
import json
import re
import warnings
from typing import Tuple, Type


def string_to_np_dtype(dtype: str) -> Type[np.dtype]:
    if dtype == 'u8' or dtype == 'uint8':
        return np.uint8
    elif dtype == 'u16' or dtype == 'uint16':
        return np.uint16
    elif dtype == 'u32' or dtype == 'uint32':
        return np.uint32
    elif dtype == 'u64' or dtype == 'uint64':
        return np.uint64
    elif dtype == 'i8' or dtype == 'int8':
        return np.int8
    elif dtype == 'i16' or dtype == 'int16':
        return np.int16
    elif dtype == 'i32' or dtype == 'int32':
        return np.int32
    elif dtype == 'i64' or dtype == 'int64':
        return np.int64
    elif dtype == 'dbl' or dtype == 'double' or dtype == 'float':
        return np.double
    elif dtype == 'sgl' or dtype == 'single':
        return np.single
    else:
        warnings.warn(f'dtype string "{dtype}" is invalid or empty; defaulting to {np.uint32}')
        return np.uint32


def downscale_u32_u8(image: np.ndarray, factor: float = 1.0) -> np.ndarray:
    image = image.astype(np.double)
    image = np.multiply(np.divide(image, np.max(image)*factor), 255)
    return image.astype(np.uint8)


def dB_u32_to_u8(image: np.ndarray, floor: float = 50) -> np.ndarray:
    image = np.multiply(np.log10(np.divide(image.astype(np.double), np.max(image))), 10)
    image = np.multiply(np.divide(image+floor, np.max(image+floor)), 255)
    return image.astype(np.uint8)


class ocmbin:
    __HEADER_SIZE__ = 8192
    __HEADER_NULL__ = re.compile('\x00')

    def __init__(self, path: str, do_load: True):
        pass

        self.path = path

        if not hasattr(self, 'is_loaded'):
            self.is_loaded = False

            # Metadata
            self.metadata = None
            self.datasets = None
            self.overlap = None

            # Data
            self.cube = None
            self.cube_reduced = None
            self.position = None
            self.data = None

        if do_load:
            self.load_data()

    def load_data(self):
        if not self.is_loaded:
            if self.metadata is None:
                self.parse_metadata()
                self.check_overlap()

            self.extract_data()

            self.is_loaded = True

    def unload_data(self):
        if self.is_loaded:
            # Release data, keep metadata
            self.cube = None
            self.cube_reduced = None
            self.position = None
            self.data = None

            self.is_loaded = True

    def read_header(self):
        with open(self.path, 'rb') as f:
            f.seek(0,0)
            header = f.read(self.__HEADER_SIZE__)
            header = header.decode('UTF-8').strip()
            header = self.__HEADER_NULL__.sub('', header)
            return header

    def parse_metadata(self):
        header = self.read_header()
        self.metadata = json.JSONDecoder().decode(header)

        self.datasets = [dataset['Name'] for dataset in self.metadata['Data']]

    def check_overlap(self):
        pass

        self.overlap = {}

        for i, dataset in enumerate(self.datasets):
            self.overlap[dataset] = self.metadata['Data'][i]['Position']['LastByte'] > \
                                    self.metadata['Data'][i]['Position']['StopByte']

            self.overlap[dataset] = {
                'overbyte': self.metadata['Data'][i]['Position']['LastByte'] -
                            self.metadata['Data'][i]['Position']['StopByte'],
                'allocated':    self.metadata['Data'][i]['Position']['StopByte'] - \
                                self.metadata['Data'][i]['Position']['StartByte'],
                'written':      self.metadata['Data'][i]['Position']['LastByte'] - \
                                self.metadata['Data'][i]['Position']['StartByte'],

            }

    def extract_data(self, load_mode: str = 'read'):
        self.data = {}

        # Extract all datasets
        for i, dataset in enumerate(self.datasets):
            self.data[dataset] = self.read_region(
                self.metadata['Data'][i]['Position']['StartByte'],
                tuple(self.metadata['Data'][i]['Size']),
                load_mode,
                string_to_np_dtype(self.metadata['Data'][i]['dtype'])
            )

        # Handle main datasets, 'cube' and 'position'
        if 'cube' in self.data.keys():
            self.cube = self.data['cube']

        if 'position' in self.data.keys():
            self.position = self.data['position']

            # Convert uint32 nm to float µm
            self.position = self.position.astype(np.double) / 1e3

    def read_region(self, start: int, shape: Tuple[int, int, int],
                    load_mode: str = 'read', dtype: Type[np.dtype] = np.uint32) -> np.ndarray:
        dimensions = len(shape)
        count = int(np.prod(shape))

        if load_mode == 'memmap':
            # Memmap region of binary file
            raise NotImplementedError
        else:
            # Read into memory
            with open(self.path, 'rb') as f:
                f.seek(start, 0)  # Go to 'StartByte'
                data = np.fromfile(f, dtype=dtype, count=count)

                if dimensions == 1:
                    return data
                elif dimensions == 2:
                    data = data.reshape(shape)
                    return data.transpose()
                elif dimensions == 3:
                    fixed_shape = (shape[2], shape[1], shape[0])
                    data = data.reshape(fixed_shape).newbyteorder('<')
                    return data.transpose((2, 1, 0))  # Transpose to X*Y*Z  todo: double-check!
                else:
                    raise NotImplementedError(f'Invalid dataset shape: {shape}; must be either 1-, 2-, or 3-dimensional')
